# Task
Below you can see two datasets Write the code to preprocess Absenteeism_data.csv to Absenteeism_preprocessed.csv. use markdown cells to define and explain each step

Here is all the data you need:
"Absenteeism_data.csv"
"Absenteeism_preprocessed.csv"

## Data loading

### Subtask:
Load the "Absenteeism_data.csv" file into a pandas DataFrame.


**Reasoning**:
Load the "Absenteeism_data.csv" file into a pandas DataFrame and display its first few rows and information.



In [1]:
import pandas as pd

try:
    df = pd.read_csv('Absenteeism_data.csv')
    display(df.head())
    display(df.info())
except FileNotFoundError:
    print("Error: 'Absenteeism_data.csv' not found. Please ensure the file exists in the current directory.")


ID  Reason for Absence        Date  Transportation Expense  \
0  11                  26  07/07/2015                     289   
1  36                   0  14/07/2015                     118   
2   3                  23  15/07/2015                     179   
3   7                   7  16/07/2015                     279   
4  11                  23  23/07/2015                     289   

   Distance to Work  Age  Daily Work Load Average  Body Mass Index  Education  \
0                36   33                  239.554               30          1   
1                13   50                  239.554               31          1   
2                51   38                  239.554               31          1   
3                 5   39                  239.554               24          1   
4                36   33                  239.554               30          1   

   Children  Pets  Absenteeism Time in Hours  
0         2     1                          4  
1         1     0                          0  
2         0     0                          2  
3         2     0                          4  
4         2     1                          2

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 700 entries, 0 to 699
Data columns (total 12 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   ID                         700 non-null    int64  
 1   Reason for Absence         700 non-null    int64  
 2   Date                       700 non-null    object 
 3   Transportation Expense     700 non-null    int64  
 4   Distance to Work           700 non-null    int64  
 5   Age                        700 non-null    int64  
 6   Daily Work Load Average    700 non-null    float64
 7   Body Mass Index            700 non-null    int64  
 8   Education                  700 non-null    int64  
 9   Children                   700 non-null    int64  
 10  Pets                       700 non-null    int64  
 11  Absenteeism Time in Hours  700 non-null    int64  
dtypes: float64(1), int64(10), object(1)
memory usage: 65.8+ KB


None

## Data exploration

### Subtask:
Explore the loaded dataset to understand its structure, identify potential issues like missing values, and examine data types.  Pay close attention to the 'Reason for Absence' column.


**Reasoning**:
Explore the dataset by examining its shape, data types, missing values, and descriptive statistics, focusing on the 'Reason for Absence' column and the 'Date' column.



In [2]:
# Examine the shape of the DataFrame
print("Shape of the DataFrame:", df.shape)

# Re-examine data types and check for null values; investigate 'Reason for Absence'
print("\nDataFrame Information:")
display(df.info())
print("\nValue counts for 'Reason for Absence':")
display(df['Reason for Absence'].value_counts())

# Descriptive statistics for numerical features
print("\nDescriptive Statistics:")
display(df.describe())

# Investigate the 'Date' column
print("\nInformation about the 'Date' column:")
print("Data type of 'Date' column:", df['Date'].dtype)
# Convert 'Date' column to datetime objects
try:
    df['Date'] = pd.to_datetime(df['Date'], format='%d/%m/%Y')
    print("Successfully converted 'Date' column to datetime objects.")
except ValueError as e:
    print(f"Error converting 'Date' column to datetime: {e}")

# Explore the distribution of 'Absenteeism Time in Hours'
print("\nDistribution of 'Absenteeism Time in Hours':")
display(df['Absenteeism Time in Hours'].describe())

Shape of the DataFrame: (700, 12)

DataFrame Information:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 700 entries, 0 to 699
Data columns (total 12 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   ID                         700 non-null    int64  
 1   Reason for Absence         700 non-null    int64  
 2   Date                       700 non-null    object 
 3   Transportation Expense     700 non-null    int64  
 4   Distance to Work           700 non-null    int64  
 5   Age                        700 non-null    int64  
 6   Daily Work Load Average    700 non-null    float64
 7   Body Mass Index            700 non-null    int64  
 8   Education                  700 non-null    int64  
 9   Children                   700 non-null    int64  
 10  Pets                       700 non-null    int64  
 11  Absenteeism Time in Hours  700 non-null    int64  
dtypes: float64(1), int64(10), object(1)
memory usage

None


Value counts for 'Reason for Absence':


Reason for Absence
23    147
28    110
27     66
13     52
0      38
19     36
22     32
26     31
25     29
11     24
10     22
18     21
14     18
1      16
7      13
12      8
21      6
6       6
8       5
9       4
5       3
24      3
16      3
15      2
4       2
17      1
3       1
2       1
Name: count, dtype: int64


Descriptive Statistics:


ID  Reason for Absence  Transportation Expense  \
count  700.000000          700.000000              700.000000   
mean    17.951429           19.411429              222.347143   
std     11.028144            8.356292               66.312960   
min      1.000000            0.000000              118.000000   
25%      9.000000           13.000000              179.000000   
50%     18.000000           23.000000              225.000000   
75%     28.000000           27.000000              260.000000   
max     36.000000           28.000000              388.000000   

       Distance to Work         Age  Daily Work Load Average  Body Mass Index  \
count        700.000000  700.000000               700.000000       700.000000   
mean          29.892857   36.417143               271.801774        26.737143   
std           14.804446    6.379083                40.021804         4.254701   
min            5.000000   27.000000               205.917000        19.000000   
25%           16.000000   31.000000               241.476000        24.000000   
50%           26.000000   37.000000               264.249000        25.000000   
75%           50.000000   40.000000               294.217000        31.000000   
max           52.000000   58.000000               378.884000        38.000000   

        Education    Children        Pets  Absenteeism Time in Hours  
count  700.000000  700.000000  700.000000                 700.000000  
mean     1.282857    1.021429    0.687143                   6.761429  
std      0.668090    1.112215    1.166095                  12.670082  
min      1.000000    0.000000    0.000000                   0.000000  
25%      1.000000    0.000000    0.000000                   2.000000  
50%      1.000000    1.000000    0.000000                   3.000000  
75%      1.000000    2.000000    1.000000                   8.000000  
max      4.000000    4.000000    8.000000                 120.000000


Information about the 'Date' column:
Data type of 'Date' column: object
Successfully converted 'Date' column to datetime objects.

Distribution of 'Absenteeism Time in Hours':


count    700.000000
mean       6.761429
std       12.670082
min        0.000000
25%        2.000000
50%        3.000000
75%        8.000000
max      120.000000
Name: Absenteeism Time in Hours, dtype: float64

## Data cleaning

### Subtask:
Check for and handle any inconsistencies or duplicates in the data.  Specifically, investigate the `Reason for Absence` column further, and address any issues related to it.  Also ensure that the data types of all columns are correct and consistent with the expected data.


**Reasoning**:
Investigate the 'Reason for Absence' column for inconsistencies, check for duplicates, verify data types, and optionally handle outliers.



In [3]:
# Investigate 'Reason for Absence'
print("Unique values in 'Reason for Absence':", df['Reason for Absence'].unique())
print("Number of unique values:", df['Reason for Absence'].nunique())

# Check for duplicates
duplicates = df.duplicated()
num_duplicates = duplicates.sum()
print(f"\nNumber of duplicate rows: {num_duplicates}")
if num_duplicates > 0:
    df = df.drop_duplicates()
    print("Duplicate rows removed.")

# Verify Data Types
print("\nData Types before correction:")
display(df.dtypes)

# Check if 'Reason for Absence' needs to be categorical
if pd.api.types.is_numeric_dtype(df['Reason for Absence']):
    df['Reason for Absence'] = df['Reason for Absence'].astype('object')
    print("'Reason for Absence' converted to object type.")

print("\nData Types after correction:")
display(df.dtypes)


# Address Outliers (Optional):
# For this example, we will cap 'Absenteeism Time in Hours' at 3 standard deviations above the mean
# This is just one approach. Other methods could include winsorization or removing outliers.
upper_limit = df['Absenteeism Time in Hours'].mean() + 3 * df['Absenteeism Time in Hours'].std()
df['Absenteeism Time in Hours'] = df['Absenteeism Time in Hours'].clip(upper=upper_limit)
print(f"\n'Absenteeism Time in Hours' capped at {upper_limit:.2f}")


Unique values in 'Reason for Absence': [26  0 23  7 22 19  1 11 14 21 10 13 28 18 25 24  6 27 17  8 12  5  9 15
  4  3  2 16]
Number of unique values: 28

Number of duplicate rows: 8
Duplicate rows removed.

Data Types before correction:


ID                                    int64
Reason for Absence                    int64
Date                         datetime64[ns]
Transportation Expense                int64
Distance to Work                      int64
Age                                   int64
Daily Work Load Average             float64
Body Mass Index                       int64
Education                             int64
Children                              int64
Pets                                  int64
Absenteeism Time in Hours             int64
dtype: object

'Reason for Absence' converted to object type.

Data Types after correction:


<ipython-input-3-cee0b0a01080>:19: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['Reason for Absence'] = df['Reason for Absence'].astype('object')


ID                                    int64
Reason for Absence                   object
Date                         datetime64[ns]
Transportation Expense                int64
Distance to Work                      int64
Age                                   int64
Daily Work Load Average             float64
Body Mass Index                       int64
Education                             int64
Children                              int64
Pets                                  int64
Absenteeism Time in Hours             int64
dtype: object


'Absenteeism Time in Hours' capped at 45.01


<ipython-input-3-cee0b0a01080>:30: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['Absenteeism Time in Hours'] = df['Absenteeism Time in Hours'].clip(upper=upper_limit)


## Data wrangling

### Subtask:
Transform the 'Reason for Absence' column into dummy variables, and perform other necessary data transformations.


**Reasoning**:
Transform the 'Reason for Absence' column into dummy variables, group them, drop the original column, and perform other necessary data transformations as instructed.



In [4]:
# Create dummy variables for 'Reason for Absence'
reason_columns = pd.get_dummies(df['Reason for Absence'], drop_first=True)

# Group the dummy variables into four categories
reason_type_1 = reason_columns.loc[:, 1:14].max(axis=1)
reason_type_2 = reason_columns.loc[:, 15:17].max(axis=1)
reason_type_3 = reason_columns.loc[:, 18:21].max(axis=1)
reason_type_4 = reason_columns.loc[:, 22:].max(axis=1)

# Concatenate the grouped dummy variables with the original DataFrame
df = pd.concat([df, reason_type_1, reason_type_2, reason_type_3, reason_type_4], axis=1)

# Rename the new columns
column_names = ['Date', 'Transportation Expense', 'Distance to Work', 'Age',
                'Daily Work Load Average', 'Body Mass Index', 'Education',
                'Children', 'Pets', 'Absenteeism Time in Hours', 'Reason_1', 'Reason_2', 'Reason_3', 'Reason_4']
df.columns = ['ID', 'Reason for Absence'] + column_names

# Drop the original 'Reason for Absence' column
df = df.drop(['Reason for Absence'], axis=1)
df = df.drop(['ID'], axis=1)


# Map education levels to numerical values
df['Education'] = df['Education'].map({1:0, 2:1, 3:1, 4:1})

# Extract month from the 'Date' column
df['Month Value'] = df['Date'].dt.month

# Reorder columns (optional, but good practice)
column_names_reordered = ['Reason_1', 'Reason_2', 'Reason_3', 'Reason_4', 'Month Value'] + [col for col in df.columns if col not in ['Reason_1', 'Reason_2', 'Reason_3', 'Reason_4', 'Month Value']]
df = df[column_names_reordered]


# Display the preprocessed DataFrame
display(df.head())

Reason_1  Reason_2  Reason_3  Reason_4  Month Value       Date  \
0     False     False     False      True            7 2015-07-07   
1     False     False     False     False            7 2015-07-14   
2     False     False     False      True            7 2015-07-15   
3      True     False     False     False            7 2015-07-16   
4     False     False     False      True            7 2015-07-23   

   Transportation Expense  Distance to Work  Age  Daily Work Load Average  \
0                     289                36   33                  239.554   
1                     118                13   50                  239.554   
2                     179                51   38                  239.554   
3                     279                 5   39                  239.554   
4                     289                36   33                  239.554   

   Body Mass Index  Education  Children  Pets  Absenteeism Time in Hours  
0               30          0         2     1                        4.0  
1               31          0         1     0                        0.0  
2               31          0         0     0                        2.0  
3               24          0         2     0                        4.0  
4               30          0         2     1                        2.0

## Data preparation

### Subtask:
Save the preprocessed DataFrame to a new CSV file named 'Absenteeism_preprocessed.csv'.


**Reasoning**:
Save the preprocessed DataFrame `df` to a new CSV file named 'Absenteeism_preprocessed.csv', without the index.



In [5]:
df.to_csv('Absenteeism_preprocessed.csv', index=False)

## Summary:

### 1. Q&A

No questions were explicitly asked in the provided data analysis task. However, the task implicitly asks for a preprocessed version of the "Absenteeism_data.csv" file.  The steps involved answer this implicit question by creating a new CSV file named "Absenteeism_preprocessed.csv."


### 2. Data Analysis Key Findings

*   **Duplicate Rows Removed:** Eight duplicate rows were identified and removed from the dataset.
*   **'Reason for Absence' Transformation:** The 'Reason for Absence' column, initially containing 28 unique numerical codes, was converted into four new binary columns ('Reason_1', 'Reason_2', 'Reason_3', 'Reason_4') representing grouped absence categories.  The original 'Reason for Absence' column was then dropped.
*   **'Education' Column Transformation:** The 'Education' column was recoded, mapping values 1 to 0, and 2, 3, and 4 to 1.
*   **'Month Value' Feature Created:** A new 'Month Value' column was extracted from the 'Date' column.
*   **Outlier Handling in 'Absenteeism Time in Hours':** Values exceeding three standard deviations above the mean (approximately 45.01 hours) were capped at this limit.
* **'ID' Column Dropped**: The 'ID' column was removed from the final dataset.


### 3. Insights or Next Steps

*   **Investigate Absence Reasons:** Further explore the meaning of the different absence categories (now represented by 'Reason_1' to 'Reason_4') to gain deeper insights into the patterns of absenteeism.  This might involve consulting the original data source or domain experts.
*   **Model Building:** Use the preprocessed data to build a predictive model for absenteeism. This could involve exploring different machine learning algorithms and evaluating their performance.
